keps/0017-pipe-tasks.md

```diff
@@ -0,0 +1,122 @@
+                  key: {{.Pipes.Certificate}}
+``` 
+The key `{{.Pipes.Certificate}}` value will be generated by KUDO to avoid collisions and to make sure that it is not used before the file is generated. In the above example we would generate a secret name like `deploy.bootstrap.gencert.pipes.certificate-#hash` to capture which plan/phase/step/task it belongs too as well as the hash of its content. KUDO would use this hash to avoid recreating secrets for unchanged files on a plan rerun. We would also use labels ensuring that the secret is cleaned up when the corresponding Instance is removed.
+
+The corresponding `gencert` task can only have one resource definition:
+```yaml
+tasks:
+  gencert:
+    resources:
+      - gencert.yaml
+```
+
+For the pipe task resources, we only allow a [ContainerSpec](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.10/#container-v1-core) to be specified. Reasons for that are explained below in the implementation details. The ContainerSpec has to define a shared volume where the files are stored.
+
+```yaml
+spec:
+  volumes:
+  - name: shared-data
+    emptyDir: {}
+
+  containers:
+    - name: gencert
+      image: frapsoft/openssl
+      imagePullPolicy: Always
+      command: [ "sh", "-c" ]
+      args:
+        - openssl req -new -newkey rsa:4096 -x509 -sha256 -days 365 -nodes -out MyCertificate.crt -keyout /usr/share/MyKey.key
+      volumeMounts:
+        - name: shared-data
+          mountPath: /usr/share/
+```
+
+Any subsequent task resource might reference previously generated file by its key `{{.Pipes.Certificate}}` e.g.:
+```yaml
+spec:
+  containers:
+  - name: myapp
+    image: myapp:latest
+    volumeMounts:
+    - name: cert
+      mountPath: /еtc/certs
+  volumes:
+  - name: cert
+    secret:
+      secretName: {{.Pipes.Certificate}}
+```
+
+### Limitations
+- File generating container has to be idempotent as the container might be executed multiple times on failure. If that's not the case, `restartPolicy: Never` has to be set which would prevent the task from finishing successfully.
+- Only files <1Mb are applicable to be stored as ConfigMap or Secret.
+- Only one resource per pipe task is allowed. If needed multiple tasks must be used.
+
+### Implementation Details/Notes/Constraints
+
+There are several ways to implement pipe tasks, each one having its challenges and complexities. The approach below allows us not to worry about Pod container life-cycle as well as keep the storing logic in the KUDO controller:
+- Provided ContainerSpec is injected into a Pod as an [InitContainer](https://kubernetes.io/docs/concepts/workloads/pods/init-containers/). This is the simplest way to wait for container completion. This is also the reason why pipe task resource definition is a ContainerSpec and not a complete Pod specification.
+- The main container is a `busybox` image, running the `sleep infinity` command, which purpose is to wait for KUDO to extract and store the files.
```

Comment:
Did you think about this "sidecar" actually doing the copying? it feels more natural to me and easier to debug, easier to see if something did not work out and more encapsulated... than going to controller code and figuring out if it already picked up the copying etc.

Reply:
I thought about that and we could do it like this, however the sidecar:
- would be a separate sub-project/executable/docker image to develop/maintain/release/run
- would need configuration/permissions to access cluster/store files
- at this point, it is basically the sidecar from the alternative solution minus waiting for the other container to finish
Did I mention that no code is the best code? 😄

Reply:
I am just a bit worried of the debugging experience when sidecar is waiting on something, you have to go to controller code where millions of things are happening and figure out why your copy is stuck... I guess that's my main concern around the controller solution, although I see the advantages...

Reply:
Fair but that debugging problem always exists for any plan/step/tasks that are stuck? Actually, keeping the copying/storing logic in KUDO controller hopefully makes it easier to debug since we would generate corresponding events.

Reply:
The KUDO controller could publish events or update a status somewhere that could help with this.